contracts/SnappAuction.sol

```diff
@@ -0,0 +1,81 @@
+pragma solidity ^0.5.0;
+
+import "./SnappBase.sol";
+
+
+contract SnappAuction is SnappBase {
+  
+    uint16 public constant AUCTION_BATCH_SIZE = 1000;
+
+    uint public auctionIndex;
+    mapping (uint => PendingFlux) public auctions;
+
+    event SellOrder(
+        uint auctionId, 
+        uint16 slotIndex,
+        uint16 accountId, 
+        uint8 buyToken, 
+        uint8 sellToken, 
+        uint128 buyAmount,
+        uint128 sellAmount
+    );
+
+    constructor () public {
+        auctions[auctionIndex].creationBlock = block.number;
+    }
+
+    function placeSellOrder(
+        uint8 buyToken,
+        uint8 sellToken,
+        uint128 minBuyAmount,
+        uint128 maxSellAmount
+    ) public onlyRegistered() {
+        require(buyToken != sellToken, "Buy and Sell tokens must differ!");
+
+        // Must have 0 < tokenId < MAX_TOKENS anyway, so might as well ensure registered.
+        require(tokenIdToAddressMap[buyToken] != address(0), "Buy token is not registered");
+        require(tokenIdToAddressMap[sellToken] != address(0), "Sell token is not registered");
+
+        // Restrict buy and sell amount to occupy less than 100 bits.
```

Comment:
nit: I believe this should say `to occupy at most 100 bits`

```python
>>> int('1'*100, 2) < 0x10000000000000000000000000
True
```

Reply:
All these checks and require have the disadvantage that they cost gas for all users. While it is good for users, who make mistakes, its bad for all other users. I suggest to remove all these requires...

Reply:
If someone submits an order with `tokenId` out of range, then the whole listener will break. For example, we will get IndexErrors on the balance array. Do we even know what the solver will do if the limit orders are between the same token pairs? Have we considered how to handle orders for non-existent token pairs?

I agree that this consumes gas, but where should this verification logic go? If it isn't handled at the source or error, then I presume we will have a lot of special cases to consider in the external services. 

I can remove all of this validation easily, but want to make sure we are all aware of the consequences we will have to face upon doing so.

Reply:
You are right that it would be more annoying to deal with all these special cases in the listener, and also the driver would have to do the same. It's not so bad, but still unpleasant.

 Do we even know what the solver will do if the limit orders are between the same token pairs?
-> this can be added
Have we considered how to handle orders for non-existent token pairs?
-> user balances will be zero, so there can't happen any trade anyways

Reply:
But if the tokenIds are greater than the max, we will get index errors when trying to access the balances....

Reply:
I guess our current choice of 30 tokens is not ideal. We should probably chose a power of 2 (e.g. 32) and then just trim all input so we can ensure the value is in range.

I'm also ok with leaving these requires for now.

Reply:
I will offload this discussion to the Contract Optimization Wiki

[Contract Optimization Wiki](https://github.com/gnosis/dex-contracts/wiki/Contract-Optimization-Opportunities)